### PR TITLE
add env var support into executables

### LINKED ISSUE #39840: Add env var support for executable table functions
Now we have the `SETTINGS` support for both `executable` engine and table function.
It's time to get env variables there.
Proposed syntax:
```sql
SETTINGS env.VAR1 = 'a', env.VAR2 = 1, env.VAR3 = 'true'
```
This way we can set up different views and tables using the same scripts with different settings.

Why it's better than using arguments: we can ride on parser ser/de for various type checks and `toString()` implementation.
Otherwise strings in strings need to be escaped and each script should have arg parser wrapper, etc.

Will add PoC implementation soon

### PR DESCRIPTION
**Pull request by pkit**
### UPDATED:

- adds a new setting `ENVIRONMENT VAR1 = 'a', VAR2 = 'b'` to table
functions
- improves "set query" handler to work on any settings type, even partial
- adds alow list for env vars to config (with globbing support)

Multiline config:
```xml
    <user_scripts>
        <allow_env_variables>
            <VAR_*/>
            <VAR1/>
        </allow_env_variables>
    </user_scripts>
```
Single pattern config:
```xml
    <user_scripts>
        <allow_env_variables>*</allow_env_variables>
    </user_scripts>
```

Query syntax:

```sql
SELECT * FROM
  executable('<script name>', '<format>', '<columns>',
    (SELECT * FROM table),
    ENVIRONMENT VAR1 = 'test', VAR2 = 1,
    SETTINGS <.....>
  )
```

Note: max additional environment size is limited to 4KB

### DEPRECATED:

Syntax:
```sql
SELECT * FROM
  executable('<script name>', '<format>', '<columns>',
    (SELECT * FROM table),
    SETTINGS env.VAR1 = 'test', env.VAR2 = 1
  )
```
_note: will **not** overwrite existing env vars_

fixes: #39840

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
environment variable support for executable table functions

### CODE REVIEW

**Comment by kitaisreal** [comment]
@pkit you can pass parameters into function, should we add also env var support ?

**Comment by pkit** [comment]
@kitaisreal I'm not sure how to pass params into function so it doesn't require escaping too much
Do you have an idea for a good syntax ?

**Comment by kitaisreal** [comment]
@pkit we already have such functionality, it works only for `executable` functions with `execute_direct`. Check example in docs https://clickhouse.com/docs/en/sql-reference/functions/#executable-user-defined-functions.
Script will look like this:
```python3
#!/usr/bin/python3

import sys

if __name__ == "__main__":
    for line in sys.stdin:
        print("Parameter " + str(sys.argv[1]) + " value " + str(line), end="")
        sys.stdout.flush()
```
Query:
```sql
SELECT test_function_parameter_python(1)(2);
```
Result:
```
┌─test_function_parameter_python(1)(2)─┐
│ Parameter 1 value 2                  │
└──────────────────────────────────────┘
```

**Comment by pkit** [comment]
@kitaisreal in each script we need to map position to key and it works essentially in UDF only, not in UDTF?
Dunno if it's good enough.
Env vars is a standard way to accept arguments, obviously we can generate wrapper scripts with argument setup that we need, but that's actually another exec call with all the costs...
Just consider the situation in the wild:
AWS Lambda - env var support
Cloudflare Workers - env var support

**Comment by alexey-milovidov** [comment]
It's strange to put them inside the SETTINGS clause and introduce a special case for the syntax of their names.
Maybe make it with

`SETTINGS a = b ENVIRONMENT c = d`

?

**Comment by alexey-milovidov** [comment]
Also, I suspect, this is dangerous.

We ensure that the user can only use executables specified in the configuration file.
But with env override, like LD_PRELOAD, you can do arbitrary code execution.

**Comment by alexey-milovidov** [comment]
How to avoid this:
- add allow-list of environment variables (in the UDF config) that can be changed;
- do nothing, let all the variables passed uniformly via command line arguments, and let everything else be handled by a wrapper executable, configured by a user.

**Comment by pkit** [comment]
> SETTINGS a = b ENVIRONMENT c = d

Yeah, I've also considered that one.
If it's better, I can do it.

> do nothing, let all the variables passed uniformly via command line arguments, and let everything else be handled by a wrapper executable, configured by a user.

Yup, I had this one in the first message. It means we are setting up a new process twice per each call to executable.
It's ok for a long running stuff, but for something pumped by blocks in real-time through materialized view... it looks suboptimal.
Will think about it.

What about running real tightened process when?
I.e. do something similar to what k8s or docker does for processes: clone mount points, pids, uids, uts etc. but not networking.
It should be probably another engine though.

The ultimate goal here is an ability to pump blocks in real time through some processing that does non-trivial transformation logic (DL/ML model, external service access, etc.) quite a lot of features are missing for that one and I will try to add them gradually.

**Comment by alexey-milovidov** [comment]
> but not networking.
It should be probably another engine though.

Networking is needed for some UDFs: for example, DNS requests.

Let's add configurable allowed env variables per UDF.

**Comment by pkit** [comment]
> Networking is needed for some UDFs

Here I mean that the namespace for networking will not change, as there is no point to mess with it.
Will be equivalent of `docker .... --network=host`
Essentially it's like running docker but with less restrictions, as a lot of them do not make sense for an executable that's fed from fd streams.

> Let's add configurable allowed env variables per UDF.

Can do it. Is support for globbing ok there? (i.e. `allowed_env_vars=*` or somthing)

Are you in favor of `ENVIRONMENT a=b` syntax?

**Comment by alexey-milovidov** [comment]
> Can do it. Is support for globbing ok there?

Yes.

> Are you in favor of ENVIRONMENT a=b syntax?

Yes.

**Comment by pkit** [comment]
@alexey-milovidov @kitaisreal I'm not sure where in config should I put the `allowed_env_variables` expression?
For UDF and UDict it's anyway configured from config only, but there is no specific config per UDTF?
We can allow env vars in config per UDF/UDict directly, I suppose?

**Comment by alexey-milovidov** [comment]
Yes, inside the config of every function, so it can be tuned for every function separately.

**Comment by pkit** [comment]
@alexey-milovidov I still cannot get the exact syntax, for example if the table function looks like:
```sql
SELECT * FROM
  executable('my_model.py predict', 'ArrowStream', 'prediction Float32',
    (SELECT * FROM table),
    ENVIRONMENT VAR1 = 'test', VAR2 = 1
  )
```
How exactly the entry in config file should look like?

**Comment by alexey-milovidov** [comment]
Ok, I understand - executable table functions don't have per-function config entry.
Let's do something like:

```
<user_scripts>
    <allow_env_variables>
        <var1/>
        <var2/>
    </allow_env_variables>
</user_scripts>
```

- globally allowed env variables for table functions and UDFs.

**Comment by pkit** [comment]
@kitaisreal @alexey-milovidov I'm hitting a wall here
I've tried to create a good parser for _partial settings_ by inheriting from the current "SetQuery" parser
```C++
class ASTSettings : public ASTSetQuery
{
public:
    explicit ASTSettings(const char * token_)
    {
        token = String(token_);
    }

    String getID(char delimiter) const override { return String("Set") + delimiter + token; }

    void formatImpl(const FormatSettings & format, FormatState &, FormatStateStacked) const override;

    String getToken() { return token; }
protected:
    String token;
};
```
And it works well in tests, everything is fine.
But when I get to actual `void TableFunctionExecutable::parseArguments(const ASTPtr & ast_function, ContextPtr context)` code the `args[i]->as<ASTSettings>()` is false.
**Does it mean that somewhere AST tree is reassembled again? And then the subclass information is lost...**

Simple test like that works fine (prints true):
```c++
int main(int, char **)
{
    using namespace DB;

    std::string input = "SELECT * FROM executable('arrow_sum.py', 'ArrowStream', 'sum UInt64', (SELECT * FROM numbers(10*1000000)), ENVIRONMENT VAR1 = 'a', SETTINGS command_termination_timeout = 100)";
    ParserQueryWithOutput parser(input.data() + input.size());
    ASTPtr ast = parseQuery(parser, input.data(), input.data() + input.size(), "", 0, 0);
    auto child = ast
                     ->children.at(0)
                     ->children.at(0)
                     ->children.at(1)
                     ->children.at(0)
                     ->children.at(0)
                     ->children.at(0);
     std::cout << !!child->as<ASTFunction>()->arguments->children.at(4)->as<ASTSettings>() << std::endl;
     return 0;
}
```

**Comment by pkit** [comment]
Ah, found it! `ASTPtr clone() const override { return std::make_shared<ASTSetQuery>(*this); }` clone always called the parent...
Indeed the tree gets reassembled through cloning...

**Comment by pkit** [comment]
OK, looks like it's working as expected finally, all features discussed here implemented
Will add tests soon

**Comment by pkit** [comment]
@kitaisreal @alexey-milovidov Done. Tests added, should be ready for review.

**Comment by alexey-milovidov** [comment]
It looks like this PR was forgotten by @kitaisreal. Now it is ready for review.

**Comment by alexey-milovidov** [comment]
This is a good PR, but it looks abandoned, and it has conflicts.

**Comment by alexey-milovidov** [comment]
Please reopen when it is ready.

File: src/Common/ShellCommand.cpp [hunk 1]
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
         /// NOTE: No escaping is performed.
         args << "'" << argv[i] << "'";
     }
+    if (!env->empty())
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:119]: `env` was passed by pointer, but wasn't checked for nullptr.

File: src/Common/ShellCommand.cpp [hunk 2]
```diff
@@ -104,7 +105,7 @@ bool ShellCommand::tryWaitProcessWithTimeout(size_t timeout_in_seconds)
     return waitForPid(pid, timeout_in_seconds);
 }
 
-void ShellCommand::logCommand(const char * filename, char * const argv[])
+void ShellCommand::logCommand(const char * filename, char * const argv[], const std::vector<std::string> *env)
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:108]: There is some inconsistency in the arguments of this function.

File: src/Common/ShellCommand.cpp [hunk 3]
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
...
         args << "'" << argv[i] << "'";
     }
+    if (!env->empty())
+    {
+        args << " (env: ";
+        for (size_t i = 0; i < env->size(); i += 2)
+        {
+            if (i > 0)
+                args << ", ";
+            args << (*env)[i] << "='" << (*env)[i + 1] << "'";
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:126]: The value isn't quoted and escaped.

File: src/Common/ShellCommand.cpp [hunk 4]
```diff
@@ -115,6 +116,17 @@ void ShellCommand::logCommand(const char * filename, char * const argv[])
         /// NOTE: No escaping is performed.
         args << "'" << argv[i] << "'";
     }
+    if (!env->empty())
+    {
+        args << " (env: ";
+        for (size_t i = 0; i < env->size(); i += 2)
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:122]: The size is evaluated every time in a loop.

File: src/Common/ShellCommand.cpp [hunk 5]
```diff
@@ -204,12 +218,17 @@ std::unique_ptr<ShellCommand> ShellCommand::executeImpl(
...
+        if (use_execve)
+            execve(filename, argv, env->data);
+        else
+            execv(filename, argv);
         /// If the process is running, then `execv` does not return here.
 
         _exit(static_cast<int>(ReturnCodes::CANNOT_EXEC));
     }
 
+    delete env;
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:230]: Oops... after reading this, I have to stop and breathe.

File: src/Common/ShellCommand.cpp [hunk 6]
```diff
@@ -335,5 +354,50 @@ void ShellCommand::wait()
...
+    if (!env_vars->empty())
+    {
+        // loosely lifted from the freebsd source
+        char ** envp;
+        for (envp = environ; envp && *envp; envp++) {}
+        size_t env_num = envp - environ;
+        orig_number_of_env_vars = env_num;
+        size_t needed_size = (env_num + env_vars->size() / 2 + 1) * sizeof(*envp);
+        data = static_cast<char **>(malloc(needed_size));
+        // here we only copy the pointers, as clickhouse should __never__ call setenv() when threads are up
```
Comment (alexey-milovidov) [inline src/Common/ShellCommand.cpp:369]: `__never__` sounds scary.

Comment (alexey-milovidov) [reply src/Common/ShellCommand.cpp:369]: `f()` is K&R style, a function with name `f` should be written as `f` in comments.

**Pull request closed without merging**